devices/src/virtio/vsock/epoll_handler.rs

```diff
@@ -0,0 +1,125 @@
+
+use super::super::super::{DeviceEventT, Error as DeviceError};
+use super::super::queue::Queue as VirtQueue;
+use super::super::VIRTIO_MMIO_INT_VRING;
+use super::defs;
+use super::{EpollHandler, VsockBackend};
+
+pub struct VsockEpollHandler<B: VsockBackend + 'static> {
+    pub rxvq: VirtQueue,
+    pub rxvq_evt: EventFd,
+    pub txvq: VirtQueue,
+    pub txvq_evt: EventFd,
+    pub evvq: VirtQueue,
+    pub evvq_evt: EventFd,
+    pub cid: u64,
+    pub mem: GuestMemory,
+    pub interrupt_status: Arc<AtomicUsize>,
+    pub interrupt_evt: EventFd,
+    pub backend: B,
+}
+
+impl<B> VsockEpollHandler<B>
+where
+    B: VsockBackend + 'static,
+{
+    /// Signal the guest driver that we've used some virtio buffers that it had previously made
+    /// available.
+    ///
+    #[allow(dead_code)]
+    fn signal_used_queue(&self) -> result::Result<(), DeviceError> {
+        debug!("vsock: raising IRQ");
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING as usize, Ordering::SeqCst);
+        self.interrupt_evt.write(1).map_err(|e| {
+            error!("Failed to signal used queue: {:?}", e);
+            DeviceError::FailedSignalingUsedQueue(e)
+        })
+    }
+
+    /// Walk the driver-provided RX queue buffers and attempt to fill them up with any data that we
+    /// have pending.
+    ///
+    #[allow(dead_code)]
+    fn process_rx(&mut self) {
+        error!("vsock: EpollHandler::process_rx() NYI");
+    }
+
+    /// Walk the dirver-provided TX queue buffers, package them up as vsock packets, and send them to
+    /// the backend for processing.
+    ///
+    #[allow(dead_code)]
+    fn process_tx(&mut self) {
+        error!("vsock: EpollHandler::process_tx() NYI");
+    }
+}
+
+impl<B> EpollHandler for VsockEpollHandler<B>
+where
+    B: VsockBackend,
+{
+    /// Respond to a new event, coming from the main epoll loop (implemented by the VMM).
+    ///
+    fn handle_event(
+        &mut self,
+        device_event: DeviceEventT,
+        evset: epoll::Events,
+    ) -> result::Result<(), DeviceError> {
+        match device_event {
+            defs::RXQ_EVENT => {
+                debug!("vsock: RX queue event");
+                if let Err(e) = self.rxvq_evt.read() {
+                    error!("Failed to get rx queue event: {:?}", e);
+                    return Err(DeviceError::FailedReadingQueue {
+                        event_type: "rx queue event",
+                        underlying: e,
+                    });
+                }
+            }
+            defs::TXQ_EVENT => {
+                debug!("vsock: TX queue event");
+                if let Err(e) = self.txvq_evt.read() {
+                    error!("Failed to get tx queue event: {:?}", e);
+                    return Err(DeviceError::FailedReadingQueue {
+                        event_type: "tx queue event",
+                        underlying: e,
+                    });
+                }
+            }
+            defs::EVQ_EVENT => {
```

Comment:
We don't yet implement the event queue. Come to think about it, we probably won't for the foreseeable future. As per the spec, the only use of the event queue is currently to let the driver know that the CID changed (for instance, due to the VM having been live migrated). However, that doesn't apply to our device model, since, in our case, the host doesn't use CIDs, so there is no risk of address space clash.

Reply:
Uh ok. Maybe a comment would be nice? (i.e // We do not currently associate functionality to this.).